### PR TITLE
IGNITE-17986 SchemaManager recovery was broken after refactoring

### PR DESCRIPTION
**Pull request by zstan**

### CODE REVIEW

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaUtils.java [hunk 1]
```diff
@@ -45,38 +45,53 @@ public static SchemaDescriptor prepareSchemaDescriptor(int schemaVer, TableView
...
-            NamedListView<? extends ColumnView> newTblColumns
+            SchemaDescriptor newDesc
     ) {
-        ColumnMapper mapper = null;
+        // Rename not supported from standard, thus only deletion.
+        if (newDesc.keyColumns().length() != oldDesc.keyColumns().length()) {
+            Optional<Column> droppedKeyCol = Arrays.stream(oldDesc.keyColumns().columns())
+                    .filter(c -> !newDesc.column(c.schemaIndex()).name().equals(c.name()))
+                    .findAny();
+
```
Comment (denis-chudov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaUtils.java:60]: I barely know anything related to schema logic, but the old algorithm of selecting `droppedKeyCol` seems more solid to me, are you sure the new one is better? What guarantees that the key columns in new schema will have the same indexes as corresponding ones in old schema?

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java [hunk 2]
```diff
@@ -301,14 +275,7 @@ private CompletableFuture<SchemaDescriptor> tableSchema(UUID tblId, String table
...
         if (checkSchemaVersion(tblId, schemaVer)) {
-            SchemaDescriptor desc = searchSchemaByVersion(tblId, schemaVer);
-
-            if (desc == null) {
-                return getSchemaDescriptor(schemaVer, tblCfg);
-            } else {
-                fut.complete(desc);
-                return fut;
-            }
+            return getSchemaDescriptor(schemaVer, tblCfg);
```
[the remark this answers is not in the document]
Comment (zstan) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java:278]: This change is just a revert my previous (possibly erroneous) fix, of course i can change javadoc (hint me plz a bit what kind of additional info i need to append) but i pointing that i just revert previous implementation.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaUtils.java (hunk 1 above)
Comment (zstan) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaUtils.java:60]: I suppose that it absolutely correct approach, plz check how this schemaIndex`es are creating : org.apache.ignite.internal.schema.Columns#sortedCopy thus if columns have different indexes it`s occasion to rise a red flag )

**Comment by zstan** [comment]
@denis-chudov plz check my comments, thanks !

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java [hunk 3]
```diff
@@ -300,27 +274,19 @@ private CompletableFuture<SchemaDescriptor> tableSchema(UUID tblId, String table
...
-                        .whenComplete((desc, th) -> {
-                            if (th != null) {
-                                fut.completeExceptionally(th);
-                            } else {
-                                fut.complete(desc);
-                            }
-                        });
+                SchemaRegistry registry0 = registriesVv.latest().get(tblId);
+
+                fut.complete(registry0.schemaCached(schemaVer));
```
Comment (denis-chudov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java:289]: could you pls add an assertion that `registry0.schemaCached(schemaVer)` didn't return `null`?

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java [hunk 4]
```diff
@@ -331,7 +297,9 @@ private CompletableFuture<SchemaDescriptor> tableSchema(UUID tblId, String table
         if (checkSchemaVersion(tblId, schemaVer)) {
             registriesVv.removeWhenComplete(schemaListener);
 
-            return getSchemaDescriptor(schemaVer, tblCfg);
+            registry = registriesVv.latest().get(tblId);
+
+            fut.complete(registry.schemaCached(schemaVer));
```
Comment (denis-chudov) [inline modules/schema/src/main/java/org/apache/ignite/internal/schema/SchemaManager.java:302]: and also here

**Comment by zstan** [comment]
@denis-chudov all done, thanks for review

**Review by korlov42 — APPROVED** [review]

**Pull request closed without merging**